go/lexer.go

```diff
@@ -0,0 +1,126 @@
+package lexer
+
+import (
+	"unicode"
+)
+
+type TokenType string
+
+const (
+	ILLEGAL   TokenType = "ILLEGAL"
+	EOF       TokenType = "EOF"
+	IDENT     TokenType = "IDENT"
+	INT       TokenType = "INT"
+	EQUAL     TokenType = "="
+	PLUS      TokenType = "+"
+	COMMA     TokenType = ","
+	SEMICOLON TokenType = ";"
+	LPAREN    TokenType = "("
+	RPAREN    TokenType = ")"
+	LSQUIRLY  TokenType = "{"
+	RSQUIRLY  TokenType = "}"
+	FUNCTION  TokenType = "FUNCTION"
+	LET       TokenType = "LET"
+)
+
+type Token struct {
+	Type    TokenType
+	Literal string
+}
+
+func NewToken(tokenType TokenType, literal string) Token {
+	return Token{Type: tokenType, Literal: literal}
+}
+
+var Keywords = map[string]TokenType{
+	"fn":  FUNCTION,
+	"let": LET,
+}
+
+type Lexer struct {
+	input        string
+	position     int
+	readPosition int
+	ch           rune
+}
+
+func NewLexer(input string) *Lexer {
+	l := &Lexer{input: input}
+	l.readChar()
+	return l
+}
+
+func (l *Lexer) NextToken() Token {
+	l.skipWhitespace()
+
+	if token, ok := getTokenFromMap(l.ch); ok {
+		l.readChar()
+		return token
+	}
+
+	if l.ch == 0 {
+		return NewToken(EOF, "")
+	}
+
+	if unicode.IsLetter(l.ch) {
+		literal := l.readIdentifier()
+		if keywordType, ok := Keywords[literal]; ok {
+			return NewToken(keywordType, literal)
+		}
+		return NewToken(IDENT, literal)
+	}
+
+	if unicode.IsDigit(l.ch) {
+		return NewToken(INT, l.readNumber())
+	}
+
+	return NewToken(ILLEGAL, string(l.ch))
+}
+
+func getTokenFromMap(ch rune) (Token, bool) {
+	var tokenMap = map[rune]Token{
```

Comment:
This can and should be declared as a pkg scoped variable, you don't want to allocate memory for this map each time you call `getTokenFromMap`, since you don't modify it and only use it for mapping.

After you do that, the function will become obsolete too.

Reply:
Good catch, fixed.